### ISSUE TITLE
@typescript-eslint/explicit-function-return-type fails with JSDoc @returns annotation

### ISSUE THREAD

**Issue by jens-duttke** [typescript-eslint/typescript-eslint#906]
I'm using JSDoc annomations to tell TypeScript the type of arguments/return values/variables, which works just great. But with the update to `@typescript-eslint/parser` version `2.0.0` I'm getting hundreds of "Missing return type on function" errors, which is wrong, because all my functions have a return type defined in the JSDoc comment (which is used by TypeScript).

**What code were you trying to parse?**

```ts
/**
 * @returns {string}
 */
function xyz () {
	return 'xyz';
}

console.log(xyz);
```

**What did you expect to happen?**

I expected that the JSDoc `@returns` annotation is considered.

**What actually happened?**

The JSDoc `@returns` annotation is not considered.

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `2.0.0` |
| `TypeScript`                | `3.5.3` |
| `ESLint`                    | `6.2.2` |
| `node`                      | `10.5.0` |
| `npm`                       | `6.4.1` |

**Comment by scottohara** [comment]
The issue here is that, starting with 2.0.0, this rule (along with a few others) is no longer restricted to only `*.ts` and `*.tsx` files.

As mentioned in the [2.0.0 release notes](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v2.0.0):
> Removed hardcoded checks which prevented some rules from running on non-TypeScript files (i.e. files that are not *.ts/*.tsx).
Some users with mixed TS/JS codebases may now see some TypeScript-specific rules being reported against JS files.
Please use ESLint's "overrides" config to select which files to apply rules to.

Whilst I understand the reasoning behind this change (to allow for things like `*.vue` SFCs that were previously not able to be linted properly), I have to say that the developer experience resulting from this change is not ideal.

Yes, it *is possible* suppress these warnings using ESLint overrides, but stepping back a bit...triggering a  `Missing return type on function` warning by default for code written in a language that *doesn't support types* makes very little sense to me.

I can imagine lots of people being surprised by this change; as aside from the aforementioned release notes, there is no documentation (at least that I could see) that explains to new users that they need to use overrides to workaround these unresolvable warnings.

Perhaps it would have been OK if ESLint overrides were able to support some kind of "wildcard rule specifier", making it possible to disable all `@typescript-eslint/*` rules for `*.js` files in one hit, e.g.

```json
{
  "overrides": [
    {
      "files": ["*.js"],
      "rules": {
        "@typescript-eslint/*": "off"
      }
    }
  ]
}
```

At it currently stands, every time a new Typescript-specific rule is added, there is a chance that it will trigger unresolvable warnings in `*.js` files, requiring a corresponding change to the ESLint overrides.

**Comment by jens-duttke** [comment]
@bradzacher  Where can I find information about this restriction in the documentation? Everywhere it's advertised that this ESLint plugin is a much better replacement for TSLint, which is deprecated now.
But on the other hand it doesn't even support the essential functionality of TypeScript, which worked fine before with TSLint.

Wouldn't it make sense to implement that in a later version, to get correct TypeScript compatiblity, instead of simply closing this issue?

**Comment by jens-duttke** [comment]
> > on the other hand it doesn't even support the essential functionality of TypeScript
>
> What do you mean by this? I would hardly call understanding JSDoc "essential functionality of typescript".

In TypeScript, type annotations in JSDoc format are as valid as inline type annotations or declaration files.
The support for JSDoc type annotations is available since version 1.8, which was released in 2015.

> My hope is that users don't blindly upgrade from v1->v2 and assume everything "just works" - I put a chunk of time into writing pretty comprehensive release notes to make sure that I could clearly educate as to the breaking changes and the fix paths.

I've read the release notes and the whole documentation, but there was no hint, that `@typescript-eslint` has only partial TypeScript support, nor is there any hint that JSDoc type annotations are not supported and rules like the `explicit-function-return-type` show erroneously messages now.

I've specified a type for the return value, on a valid way, which is supported by TypeScript since almost 4 years, but the plugins says that I havn't specified a type.

I wouldn't call this breaking change, I would call it, introducing bugs.

Since it's called `@typescript-eslint` and not `@partial-typescript-support-for-eslint`, and even Microsoft will switch to it, I expect that it supports the same functionality of TypeScript, and not a state from 5 years ago.

It's not that I have a mixed code-base, I've explictly decided to write native JavaScript for some of my projects, and using the power of TypeScript for validation and code completion in VSCode.

Part of the validation is the linting. But now, with the switch from TSLint to ESLint it feels like a step back, instead of a step forward.

**Comment by jens-duttke** [comment]
@badzacher Thanks for this detailed explanation, and sorry if my comments sounds too critical/negative. I've created this issue to help the team behind `@typescript-eslint` to improve their product. A "We do not support JSDoc annotations." is not what I've expected.
I don't know if you are in contact with the TypeScript team, but if functionality is missing in TypeScript, maybe it would make sense to communicate that to this team, instead of simply saying to your users "Whilst they are supported internally by typescript, we do not have a mechanism for parsing/interpreting them.". That sounds like you are not really interested in making your product better.

TSLint shows the information that a rule does not work for JS files explicitly in the documentation of the specific rule. Beside that, it shows one single information if you still try to apply such a rule for JS files.
On the other hand `@typescript-eslint` suppressed such warnings for JS files before (without to inform the user that the rule is not considered) and now, after the update, it shows thousands of warnings, instead of simply indicating, that the rule is not compatible to JS files.

Like @scottohara wrote

> [...] triggering a Missing return type on function warning by default for code written in a language that doesn't support types makes very little sense [...]

At this point, to me it's not clear, which rules are working with JS files, and which rules explicitly require inline-annotations. Now, I know that `explicit-function-return-type` is not working, but what about the others?
Do I need to activate all rules and write test cases manually to figure out which rules will not work for JS files and which won't? Is every user forced to do that by his own again and again? On every update?

Maybe it would make sense to add such an information atleast to the documentation of the rules, just like it is done for TSLint.

**Comment by Skalman** [comment]
I encountered the same problem just now, in a mixed TypeScript/JavaScript project. For me, this is mostly an issue of documentation. If I understand correctly, _TypeScript ESLint_ has no intention of supporting Javascript. (In contrast to the TypeScript compiler, which _does_ support JavaScript and JSDoc.)

@jens-duttke, since you're actually working on a JavaScript project, perhaps you'll have to do with just _ESLint_ lints. You can still use the TypeScript compiler.

@bradzacher, how about documenting that you should be using `overrides` together with _TypeScript ESLint_? Here's a config that I feel could be the basis of many projects.

```jsonc
{
  "env": {
    // ...
  },
  "extends": ["eslint:recommended"],
  "overrides": [
    {
      "files": ["*.ts"],
      "plugins": ["@typescript-eslint"],
      "parser": "@typescript-eslint/parser",
      "parserOptions": {
        // ...
        "project": "./tsconfig.json"
      },
      "extends": [
        "plugin:@typescript-eslint/eslint-recommended",
        "plugin:@typescript-eslint/recommended"
        // "plugin:@typescript-eslint/recommended-requiring-type-checking"
      ],
      "rules": {
        // TypeScript specific rules config
      }
    }
  ],
  "parserOptions": {
    // ...
  },
  "rules": {
    // General rules config
  }
}
```

Could it go under the heading "How do I configure my project to use typescript-eslint?" in the main README?

**Comment by bradzacher** [comment]
I've added documentation to the relevant rules mentioning how to configure overrides. 736a074b80eafd88376a90e6595a42217d939890

**Comment by Skalman** [comment]
Now, I'm a little confused. Since you're singling this rule out, are the rest of the lints supposed to work on Javascript projects too?

**Comment by bradzacher** [comment]
they should do.
As mentioned above, there are only two rules which enforce that you use TS-specific syntax - explicit-function-return-type and explicit-member-accessibility.

These two rules are the only two that straight up will not work on pure JS code. Everything else will work in some way, depending on how good your JSDoc types are, or rather how well the types can be inferred.
